File: trunk/Source/GridSoccerSimulator/Common/Position.cs
// Copyright (c) 2009 - 2010 
//  - Sina Iravanian <sina@sinairv.com>
//  - Sahar Araghi   <sahar_araghi@aut.ac.ir>
//
// This source file(s) may be redistributed, altered and customized
// by any means PROVIDING the authors name and all copyright
// notices remain intact.
// THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED. USE IT AT YOUR OWN RISK. THE AUTHOR ACCEPTS NO
// LIABILITY FOR ANY DATA DAMAGE/LOSS THAT THIS PRODUCT MAY CAUSE.
//-----------------------------------------------------------------------

using System;

namespace GridSoccer.Common
{
    public class Position
    {
        public int Row { get; set; }
        public int Col { get; set; }

        public Position(int r, int c)
        {
            this.Row = r;
            this.Col = c;
        }

        public Position()
            : this(0, 0)
        {
        }

        public Position(Position p)
            : this(p.Row, p.Col)
        {
        }

        public void Set(int r, int c)
        {
            this.Row = r;
            this.Col = c;
        }

        public void Set(Position pos)
        {
            this.Row = pos.Row;
            this.Col = pos.Col;
        }

        public static Position operator +(Position pos1, Position pos2)
        {
            int r = pos1.Row + pos2.Row;
            int c = pos1.Col + pos2.Col;
            return new Position(r, c);
        }

        public static bool operator ==(Position pos1, Position pos2)
        {
            return pos1.Equals(pos2);
        }

        public static bool operator !=(Position pos1, Position pos2)
        {
            return !pos1.Equals(pos2);
        }

        public override bool Equals(object obj)
        {
            if (Object.ReferenceEquals(this, obj))
                return true;
            Position pos = obj as Position;
            if (this.Row == pos.Row && this.Col == pos.Col)
                return true;
            return false;
        }

        public override int GetHashCode()
        {
            return Row.GetHashCode() + Col.GetHashCode();
        }

        public override string ToString()
        {
            return String.Format("{0} {1}", Row, Col);
        }

        public Position Clone()
        {
            return new Position(this);
        }
    }
}

File: Source/CPPSampleAgent/CPPSampleClient/Position.cpp
// Copyright (c) 2009 - 2011 
//  - Sina Iravanian <sina@sinairv.com>
//  - Sahar Araghi   <sahar_araghi@aut.ac.ir>
//
// This source file(s) may be redistributed, altered and customized
// by any means PROVIDING the authors name and all copyright
// notices remain intact.
// THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED. USE IT AT YOUR OWN RISK. THE AUTHOR ACCEPTS NO
// LIABILITY FOR ANY DATA DAMAGE/LOSS THAT THIS PRODUCT MAY CAUSE.
//-----------------------------------------------------------------------

#include "Position.h"

Position::Position() 
{
	Row = 0;
	Col = 0;
}

Position::Position(int r, int c)
{
    Row = r;
    Col = c;
}


Position::Position(Position& p) 
{
	Row = p.Row;
	Col = p.Col;
}

void Position::Set(int r, int c)
{
    Row = r;
    Col = c;
}

void Position::Set(Position &pos)
{
    Row = pos.Row;
    Col = pos.Col;
}

Position Position::operator+(Position &rhs)
{
    int r = Row + rhs.Row;
    int c = Col + rhs.Col;
    return Position(r, c);
}

Position Position::operator -(Position &pos)
{
    int r = Row - pos.Row;
    int c = Col - pos.Col;
    return Position(r, c);
}

bool Position::operator ==(Position &pos)
{
    return Equals(pos);
}

bool Position::operator !=(Position &pos)
{
    return !Equals(pos);
}

bool Position::Equals(Position &pos)
{
    if (Row == pos.Row && Col == pos.Col)
        return true;
    return false;
}

std::string Position::ToString()
{
	char szBuff[256];
	sprintf(szBuff, "%d %d", Row, Col);
	return std::string(szBuff);
}

Position Position::Clone()
{
    return Position(Row, Col);
}
